[BUG] Email Rule ignores subfolders

### Description

Mail Rules are all working fine - except if I use a subfolder.
i.e. "INBOX.Archiv" for one of my Email accounts end in
`[2022-04-18 16:08:51,911] [ERROR] [paperless_mail] Rule CMRU.cmru paperless ordner: Error while processing rule: Rule CMRU.cmru paperless ordner: Folder INBOX.Archiv does not exist in account CMRU`

Of cource I tried different spellings and different folders and different Email-Accounts. For all account only the main folder is working

### Expected behavior

Mail rule should process only email in defined subfolder

### Steps to reproduce

see above

### Webserver logs

```bash
[2022-04-18 16:08:51,862] [DEBUG] [paperless_mail] Account CMRU: Processing 1 rule(s)

[2022-04-18 16:08:51,867] [DEBUG] [paperless_mail] Rule CMRU.cmru paperless ordner: Selecting folder INBOX.Archiv

[2022-04-18 16:08:51,911] [ERROR] [paperless_mail] Rule CMRU.cmru paperless ordner: Error while processing rule: Rule CMRU.cmru paperless ordner: Folder INBOX.Archiv does not exist in account CMRU

Traceback (most recent call last):

  File "/usr/src/paperless/src/paperless_mail/mail.py", line 192, in handle_mail_rule

    M.folder.set(rule.folder)

  File "/usr/local/lib/python3.9/site-packages/imap_tools/folder.py", line 44, in set

    check_command_status(result, MailboxFolderSelectError)

  File "/usr/local/lib/python3.9/site-packages/imap_tools/utils.py", line 45, in check_command_status

    raise exception(command_result=command_result, expected=expected)

imap_tools.errors.MailboxFolderSelectError: Response status "OK" expected, but "NO" received. Data: [b"INBOX.Archiv doesn't exist."]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "/usr/src/paperless/src/paperless_mail/mail.py", line 177, in handle_mail_account

    total_processed_files += self.handle_mail_rule(M, rule)

  File "/usr/src/paperless/src/paperless_mail/mail.py", line 194, in handle_mail_rule

    raise MailError(

paperless_mail.mail.MailError: Rule CMRU.cmru paperless ordner: Folder INBOX.Archiv does not exist in account CMRU
```


### Screenshots

_No response_

### Paperless-ngx version

1.6.0

### Host OS

Synology/ Docker

### Installation method

Docker

### Browser

_No response_

### Configuration changes

_No response_

### Other

_No response_

--- a/docs/usage_overview.rst
+++ b/docs/usage_overview.rst
@@ -182,11 +182,10 @@ These are as follows:
 
     When defining a mail rule with a folder, you may need to try different characters to
     define how the sub-folders are separated.  Common values include ".", "/" or "|", but
-    this varies by the mail server.  Unfortunately, this isn't a value we can determine
-    automatically.  Either check the documentation for your mail server, or check for
-    errors in the logs and try different folder separator values.  If possible,
-    the log will contain a listing of account folders, in the event of an error when
-    trying to fetch mail from a certain folder.
+    this varies by the mail server.  Check the documentation for your mail server.  In the
+    event of an error fetching mail from a certain folder, check the Paperless logs.  When
+    a folder is not located, Paperless will attempt to list all folders found in the account
+    to the Paperless logs.
 
 .. note::
 

--- a/docs/usage_overview.rst
+++ b/docs/usage_overview.rst
@@ -184,7 +184,9 @@ These are as follows:
     define how the sub-folders are separated.  Common values include ".", "/" or "|", but
     this varies by the mail server.  Unfortunately, this isn't a value we can determine
     automatically.  Either check the documentation for your mail server, or check for
-    errors in the logs and try different folder separator values.
+    errors in the logs and try different folder separator values.  If possible,
+    the log will contain a listing of account folders, in the event of an error when
+    trying to fetch mail from a certain folder.
 
 .. note::
 

--- a/src/paperless_mail/mail.py
+++ b/src/paperless_mail/mail.py
@@ -197,7 +197,7 @@ class MailAccountHandler(LoggingMixin):
             )
             try:
                 for folder_info in M.folder.list():
-                    self.log("info", f"Located folder: {str(folder_info)}")
+                    self.log("info", f"Located folder: {folder_info.name}")
             except Exception:
                 self.log(
                     "error",

--- a/src/paperless_mail/mail.py
+++ b/src/paperless_mail/mail.py
@@ -198,10 +198,11 @@ class MailAccountHandler(LoggingMixin):
             try:
                 for folder_info in M.folder.list():
                     self.log("info", f"Located folder: {folder_info.name}")
-            except Exception:
+            except Exception as e:
                 self.log(
                     "error",
-                    "Exception during folder listing, unable to provide list folders",
+                    "Exception during folder listing, unable to provide list folders: "
+                    + str(e),
                 )
 
             raise MailError(

--- a/src/paperless_mail/mail.py
+++ b/src/paperless_mail/mail.py
@@ -190,6 +190,20 @@ class MailAccountHandler(LoggingMixin):
         try:
             M.folder.set(rule.folder)
         except MailboxFolderSelectError:
+
+            self.log(
+                "error",
+                f"Unable to access folder {rule.folder}, attempting folder listing",
+            )
+            try:
+                for folder_info in M.folder.list():
+                    self.log("info", f"Located folder: {str(folder_info)}")
+            except Exception:
+                self.log(
+                    "error",
+                    "Exception during folder listing, unable to provide list folders",
+                )
+
             raise MailError(
                 f"Rule {rule}: Folder {rule.folder} "
                 f"does not exist in account {rule.account}",

--- a/src/paperless_mail/tests/test_mail.py
+++ b/src/paperless_mail/tests/test_mail.py
@@ -15,6 +15,7 @@ from django.test import TestCase
 from documents.models import Correspondent
 from documents.tests.utils import DirectoriesMixin
 from imap_tools import EmailAddress
+from imap_tools import FolderInfo
 from imap_tools import MailboxFolderSelectError
 from imap_tools import MailMessage
 from imap_tools import MailMessageFlags
@@ -53,12 +54,11 @@ class BogusMailBox(ContextManager):
     def __init__(self):
         self.messages: List[MailMessage] = []
         self.messages_spam: List[MailMessage] = []
+        self.folder = BogusFolderManager()
 
     def login(self, username, password):
         if not (username == "admin" and password == "secret"):
             raise Exception()
-
-    folder = BogusFolderManager()
 
     def fetch(self, criteria, mark_seen, charset=""):
         msg = self.messages
@@ -620,6 +620,72 @@ class TestMail(DirectoriesMixin, TestCase):
         self.assertEqual(self.async_task.call_count, 1)
         self.assertEqual(len(self.bogus_mailbox.messages), 2)
         self.assertEqual(len(self.bogus_mailbox.messages_spam), 1)
+
+    def test_error_folder_set(self):
+        """
+        GIVEN:
+            - Mail rule with non-existent folder
+        THEN:
+            - Should call list to output all folders in the account
+            - Should not process any messages
+        """
+        account = MailAccount.objects.create(
+            name="test2",
+            imap_server="",
+            username="admin",
+            password="secret",
+        )
+        _ = MailRule.objects.create(
+            name="testrule",
+            account=account,
+            action=MailRule.AttachmentAction.MOVE,
+            action_parameter="spam",
+            filter_subject="Claim",
+            order=1,
+            folder="uuuhhhh",  # Invalid folder name
+        )
+
+        self.bogus_mailbox.folder.list = mock.Mock(
+            return_value=[FolderInfo("SomeFoldername", "|", ())],
+        )
+
+        self.mail_account_handler.handle_mail_account(account)
+
+        self.bogus_mailbox.folder.list.assert_called_once()
+        self.assertEqual(self.async_task.call_count, 0)
+
+    def test_error_folder_set_error_listing(self):
+        """
+        GIVEN:
+            - Mail rule with non-existent folder
+            - Mail account folder listing raises exception
+        THEN:
+            - Should not process any messages
+        """
+        account = MailAccount.objects.create(
+            name="test2",
+            imap_server="",
+            username="admin",
+            password="secret",
+        )
+        _ = MailRule.objects.create(
+            name="testrule",
+            account=account,
+            action=MailRule.AttachmentAction.MOVE,
+            action_parameter="spam",
+            filter_subject="Claim",
+            order=1,
+            folder="uuuhhhh",  # Invalid folder name
+        )
+
+        self.bogus_mailbox.folder.list = mock.Mock(
+            side_effect=MailboxFolderSelectError(None, "uhm"),
+        )
+
+        self.mail_account_handler.handle_mail_account(account)
+
+        self.bogus_mailbox.folder.list.assert_called_once()
+        self.assertEqual(self.async_task.call_count, 0)
 
     @mock.patch("paperless_mail.mail.MailAccountHandler.get_correspondent")
     def test_error_skip_mail(self, m):

